### PR TITLE
[WIP][ptd][nccl] use current-stream as nccl-stream under async=False mode

### LINKED ISSUE #147729: [RFE][Distributed][NCCL] A feature request for stream management API in PG NCCL
### A feature request for stream management API in PG NCCL
Asynchronous communication offers the benefit of being overlapped with other CUDA operations, thanks to stream concurrency. However, in the current state of PyTorch, this advantage may be compromised by a potential "read-before-write" issue, as each NCCL process group operates on its own dedicated stream, which may act independently.

Here is an example of a proper behavior:
```python
dist.reduce_scatter_tensor(buffer, data, group=local_pg)
dist.all_reduce(buffer, group=cross_pg)
dist.all_gather_into_tensor(data, buffer, group=local_pg)
```
Image
Where the collective ops are executed in the right order.

However, once these ops are requested as async:
```python
dist.reduce_scatter_tensor(buffer, data, group=local_pg, async_op=True)
dist.all_reduce(buffer, group=cross_pg, async_op=True)
dist.all_gather_into_tensor(data, buffer, group=local_pg, async_op=True)
```
Image
The collective ops are no longer in proper order since they run on different streams that are asynchronous wrt each other. **This behavior is certainly not what users would expect and can be considered as a bug.**

One can mitigate this with `handle.wait()` call, but the main goal is to run these comms parallel to the compute ops (matmuls etc.). We would like to propose one of two ways of setting streams in process groups:

- Pass a user defined stream as PG-option in new_group:
```python
nccl_options = dist.ProcessGroupNCCL.Options(stream=user_stream)
pg = dist.new_group(backend="nccl", pg_options=nccl_options)
```

- Set a user stream at runtime:
```python
pg1.set_stream(user_stream)
pg2.set_stream(user_stream)
dist.reduce_scatter_tensor(buffer, data, group=pg1, async_op=True)
dist.all_reduce(buffer, group=pg2, async_op=True)
```

### Motivation
It was first discovered in an attempt to do multi data center training. In the above example
```python
dist.reduce_scatter_tensor(buffer, data, group=local_pg)
dist.all_reduce(buffer, group=cross_pg)
dist.all_gather_into_tensor(data, buffer, group=local_pg)
```
The `cross_pg group` has “netName” set to “Socket” in ProcessGroupOptionsNCCL to tell NCCL to use a frontend TCP network which connects two compute clusters together, instead of IB or RoCE, that are set for `local_pg`. Without more control over the communication stream, the inter DC communication must be exposed and not overlapped with any compute.

There are many other cases that could potentially benefit from fine control of compute stream as well.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @ptrblck @msaroufim @eqy @skyw @alpha0422

### PR DESCRIPTION
**Pull request by cenzhaometa**
Summary:
PTD current workflow:
- PTD creates its own dedicated `ncclStream` for comm operation
- it will first add a dependency on current-stream (typically the compute stream) to ensure tensors are ready before invoking collective

such stream synchronization become expensive in Inference world (cpu overhead: 70us vs GPU kernel time: 160us).

This diff:
- introduces a new env `TORCH_NCCL_USE_CURRENT_STREAM_AS_NCCL_STREAM=1`
- when it's specified, PTD uses current-stream as the nccl-stream and avoids stream sync

this helps shave off 50% CPU overhead **(70us -> 35us)**, which reduce total CPU/GPU from **230us to 195us by 15%**

Test Plan:
# before
```
[cenzhao@devgpu039.atn3 ~/fbsource/fbcode (2265d32f0)]$ buck2 run @//mode/opt-amd-gpu -c fbcode.split-dwarf=True //param_bench/train/comms/pt:launcher -- --launcher mpi --nnode 1 --collective all_reduce --b 20M --e 20M --data-type bfloat16 --backend nccl --n 100 --w 5 --envs "NCCL_DEBUG_FILE=/tmp/dedicated_log_rccl.%h.%p.log;NCCL_DEBUG=INFO;NCCL_DEBUG_SUBSYS=INIT,COLL;MSCCL_ALGO_DIR=/data/users/${USER}/fbsource/third-party/rccl/develop/tools/msccl-algorithms;RCCL_MSCCLPP_THRESHOLD=$((128*1024*1024));RCCL_MSCCLPP_ENABLE=1;TORCH_NCCL_USE_TENSOR_REGISTER_ALLOCATOR_HOOK=1;" --size-start-profiler 20M
```
https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree/traces/dynocli/devgpu039.atn3.facebook.com/rank-0.Feb_24_16_19_28.354787.pt.trace.json.gz&bucket=hpc_traces

 {F1975408857} 
- c10d::allreduce_(69us)
  - cudaStreamSync (23us)
  - nccl::all_reduce(26us)

# after
```
[cenzhao@devgpu039.atn3 ~/fbsource/fbcode (2265d32f0)]$ buck2 run @//mode/opt-amd-gpu -c fbcode.split-dwarf=True //param_bench/train/comms/pt:launcher -- --launcher mpi --nnode 1 --collective all_reduce --b 20M --e 20M --data-type bfloat16 --backend nccl --n 100 --w 5 --envs "NCCL_DEBUG_FILE=/tmp/dedicated_log_rccl.%h.%p.log;NCCL_DEBUG=INFO;NCCL_DEBUG_SUBSYS=INIT,COLL;MSCCL_ALGO_DIR=/data/users/${USER}/fbsource/third-party/rccl/develop/tools/msccl-algorithms;RCCL_MSCCLPP_THRESHOLD=$((128*1024*1024));RCCL_MSCCLPP_ENABLE=1;TORCH_NCCL_USE_TENSOR_REGISTER_ALLOCATOR_HOOK=1;TORCH_NCCL_USE_CURRENT_STREAM_AS_NCCL_STREAM=1" --size-start-profiler 20M
```
https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree/traces/dynocli/devgpu039.atn3.facebook.com/rank-4.Feb_24_16_22_56.534269.pt.trace.json.gz&bucket=hpc_traces

 {F1975408962} 

- c10d:allreduce_(37us)
  - cudaStreamSync (gone)
  - nccl::all_reduce(20us)

Differential Revision: D70135605

Resolves #147729

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/_C/_distributed_c10d.pyi [hunk 1]
```diff
@@ -134,6 +134,8 @@ class BroadcastOptions:
 class AllreduceOptions:
     reduceOp: ReduceOp
     timeout: timedelta
+    asyncOp: bool
+    sparseIndices: Optional[Tensor]
```
Comment (wconstab) [inline torch/_C/_distributed_c10d.pyi:138]: Is this line added by mistake?

**Review by wconstab — COMMENTED** [review]
Lgtm except the sparseIndices line

**Comment by skyw** [comment]
Does "current stream" also include stream context under the with clause? Which sets "current stream" under the context manager.

**Comment by ngimel** [comment]
Generally, relying on env variables to control behavior is very coarse grained and not particularly convenient (e.g. when spawning processes the environment has to be passed together with processes). Can we make this a property of pg?
